src/Compilers/Core/Portable/Compilation/DeterministicKeyBuilder.cs

```diff
@@ -0,0 +1,478 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.IO;
+using System.Reflection;
+using System.Runtime.InteropServices;
+using System.Text;
+using System.Globalization;
+using Microsoft.CodeAnalysis.Debugging;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Emit;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis
+{
+    /// <summary>
+    /// The string returned from this function represents the inputs to the compiler which impact determinism.  It is 
+    /// meant to be inline with the specification here:
+    /// 
+    ///     - https://github.com/dotnet/roslyn/blob/main/docs/compilers/Deterministic%20Inputs.md
+    /// 
+    /// Issue #8193 tracks filling this out to the full specification. 
+    /// 
+    ///     https://github.com/dotnet/roslyn/issues/8193
+    /// </summary>
+    /// <remarks>
+    /// Options which can cause compilation failure, but doesn't impact the result of a successful
+    /// compilation should be included. That is because it is interesting to describe error states
+    /// not just success states. Think about caching build failures as well as build successes.
+    ///
+    /// When an option is omitted, say if there is no value for a public crypto key, we should emit
+    /// the property with a null value vs. omitting the property. Either approach would produce 
+    /// correct results the preference is to be declarative that an option is omitted.
+    /// </remarks>
+    internal abstract class DeterministicKeyBuilder
+    {
+        protected DeterministicKeyBuilder()
+        {
+        }
+
+        protected void WriteFileName(JsonWriter writer, string name, string? filePath, DeterministicKeyOptions options)
+        {
+            if ((options & DeterministicKeyOptions.IgnorePaths) != 0)
+            {
+                filePath = Path.GetFileName(filePath);
```

Comment:
these apis can throw.  we should doc at the top level that this can be possible.

Reply:
Leaving this open until we decide on the actual entry points and will add the docs then.

Reply:
Added  a comment about throwing to the entry point doc.